### ISSUE TITLE
networkd DHCPv6 client fails to process large DHCPv6 packets

### ISSUE THREAD

**Issue by nmeyerhans** [systemd/systemd#24851]
### systemd version the issue has been seen with

251

### Used distribution

Debian

### Linux kernel version used

5.19.0-2-cloud-amd64

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

Configuration of IPv6 address and route state via DHCPv6 on a network interface in response to a DHCPv6 Advertise packet containing a large number of options.

### Unexpected behaviour you saw

Systemd-networkd fails to complete the DHCPv6 negotiation and never configures addresses or routes on the interface.

The problem _may_ be related to the size of the packet payload, as it's been observed that the IP payload size is >512 bytes in failure cases, while payload sizes <512 bytes seem to work.

### Steps to reproduce the problem

The problem has been observed in an Amazon Web Services VPC network environment in which a large number (approximately >15) number of IPv6 Delegated Prefixes have been associated with a network interface. In order to reproduce this, you'll need to run a fairly large instance (e.g. c6i.16xlarge), and there's a cost associated with this.

Create a network interface with 30 delegated prefixes associated with it, e.g.

```plain
$ aws ec2 create-network-interface \
--subnet-id subnet-047cfed18eEXAMPLE \
--ipv6-address-count 1 \
--ipv6-prefix-count 30
```

Launch an instance the current Debian unstable AMI. You can obtain the AMI ID with a command similar to the following (adjusted for region, etc):
```plain
$ aws ec2 describe-images --owner 903794441882 --region us-west-2 --output json \
  --query "Images[?Architecture=='x86_64'] | [?starts_with(Name, 'debian-sid-')] | max_by([], &CreationDate)"
```
Install a .network file for the previously created ENI on the instance (adjust the MAC address to match the one from your ENI):

```plain
[Match]
Driver=ena
MACAddress=06:93:f5:28:17:39

[Link]
MTUBytes=9001

[Network]
DHCP=ipv6
LLMNR=no

[DHCPv6]
UseHostname=no
UseDNS=no
UseNTP=no

[Route]
Gateway=_ipv6ra

[IPv6AcceptRA]
UseDomains=no
```

Attach the interface to the instance, e.g.

```
$ aws ec2 attach-network-interface --output text --device-index 1 \
  --instance-id $iid --network-interface-id $eni_id --network-card-index 0
```

The same steps, but with a smaller number of prefixes assigned to the network interface (say 3 instead of 30) yields a successful DHCPv6 negotiation, and addresses and routes are configured as expected.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
This problem has also been observed with systemd 250.8 on the Amazon Linux 2022.

With debug output enabled for systemd-networkd, we see the following, which indicates that we've sent a Solicit message and processed an Advertise response, but we never seem to actually send a Request message.


$ journalctl -u systemd-networkd.service -b | grep ens6 | grep DHCPv6 | head -n30
Sep 28 17:12:21 ip-10-0-1-71 systemd-networkd[1095]: ens6: Requested configuring of the DHCPv6 client.
Sep 28 17:12:21 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client is configured.
Sep 28 17:12:22 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:22 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Starting in Solicit mode
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: State changed: stopped -> solicitation
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: Acquiring DHCPv6 lease on NDisc request
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Sent Solicit
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Next retransmission in 1s
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Processed Advertise message
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: State changed: solicitation -> request
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Next retransmission in 912ms
Sep 28 17:12:25 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Next retransmission in 1s
Sep 28 17:12:26 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Next retransmission in 3s
Sep 28 17:12:30 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Next retransmission in 6s
Sep 28 17:12:32 ip-10-0-1-71 systemd-networkd[1095]: ens6: Acquiring DHCPv6 lease on NDisc request
Sep 28 17:12:32 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:32 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:34 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
```

**Comment by nmeyerhans** [comment]
The DHCPv6 Advertise packet received by networkd in the above example is shown in wireshark's textual representation below. In it you can see the IA_NA and IA_PD options (one of the IA_PD options expanded for visibility, the rest are collapsed for brevity):

```
DHCPv6
    Message type: Advertise (2)
    Transaction ID: 0xccfe80
    Client Identifier
    Server Identifier
    Preference
    Identity Association for Non-temporary Address
        Option: Identity Association for Non-temporary Address (3)
        Length: 40
        IAID: 6618dd42
        T1: 70
        T2: 112
        IA Address
    Identity Association for Prefix Delegation
        Option: Identity Association for Prefix Delegation (25)
        Length: 1172
        IAID: 6618dd42
        T1: 70
        T2: 112
        IA Prefix
            Option: IA Prefix (26)
            Length: 25
            Preferred lifetime: 140
            Valid lifetime: 450
            Prefix length: 80
            Prefix address: 2600:1f14:eeb:2202:5a9f::
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
        IA Prefix
```

**Comment by nmeyerhans** [comment]
Adding some additional instrumentation to the code (this is generated with v250.8; I can regenerate from `main` if needed), we find that the problem occurs when trying to construct a Request packet in response to the server's Reply package.

```
Sep 28 16:30:42 ip-10-0-1-61.us-west-2.compute.internal systemd-networkd[2383]: ens6: DHCPv6 client: Recv Advertise
Sep 28 16:30:42 ip-10-0-1-61.us-west-2.compute.internal systemd-networkd[2383]: ens6: DHCPv6 client: In client_send_message()
Sep 28 16:30:42 ip-10-0-1-61.us-west-2.compute.internal systemd-networkd[2383]: ens6: DHCPv6 client: State is DHCP6_STATE_RENEW or DHCP6_STATE_REQUEST
Sep 28 16:30:42 ip-10-0-1-61.us-west-2.compute.internal systemd-networkd[2383]: ens6: DHCPv6 client: handling DHCP6_REQUEST_IA_PD
Sep 28 16:30:42 ip-10-0-1-61.us-west-2.compute.internal systemd-networkd[2383]: ens6: DHCPv6 client: Appending SD_DHCP6_OPTION_CLIENTID, return was -105
```
This comes from https://github.com/systemd/systemd-stable/blob/b16a4d45eb6c0b5ddb2d7ea5fbcf86c4f839c13d/src/libsystemd-network/sd-dhcp6-client.c#L882-L885, when it calls [dhcp6_option_append](https://github.com/systemd/systemd-stable/blob/b16a4d45eb6c0b5ddb2d7ea5fbcf86c4f839c13d/src/libsystemd-network/dhcp6-option.c#L223-L224)

**Comment by nmeyerhans** [comment]
This may simply be a matter of the hardcoded 512 byte buffer size at https://github.com/systemd/systemd/blob/07faa4990fcc1e80c9ef63c09eb91bb73dab19cb/src/libsystemd-network/sd-dhcp6-client.c#L704

Since the options sent in the Request packet should match those from the Advertise packet, the Request packet's size should be based on that of the Advertise packet.

**Comment by yuwata** [comment]
Fix is waiting in #24867.

**Comment by nmeyerhans** [comment]
Thanks, I'll test the fix as soon as possible!
